src/vanilla/store.ts

```diff
@@ -706,43 +691,72 @@ const buildStore = (
     sub: subscribeAtom,
     unstable_derive,
   }
-  if (import.meta.env?.MODE !== 'production') {
-    const devStore: DevStoreRev4 = {
-      // store dev methods (these are tentative and subject to change without notice)
-      dev4_get_internal_weak_map: () => ({
-        get: (atom) => {
-          const atomState = getAtomState(atom)
-          if (atomState.n === 0) {
-            // for backward compatibility
-            return undefined
-          }
-          return atomState
-        },
-      }),
-      dev4_get_mounted_atoms: () => debugMountedAtoms,
-      dev4_restore_atoms: (values) => {
-        const pending = createPending()
-        for (const [atom, value] of values) {
-          if (hasInitialValue(atom)) {
-            const atomState = getAtomState(atom)
-            const prevEpochNumber = atomState.n
-            setAtomStateValueOrPromise(atom, atomState, value)
-            mountDependencies(pending, atom, atomState)
-            if (prevEpochNumber !== atomState.n) {
-              addPendingAtom(pending, atom, atomState)
-              recomputeDependents(pending, atom, atomState)
+  return store
+}
+
+const deriveDevStore = (store: Store): Store & DevStoreRev4 => {
+  const debugMountedAtoms = new Set<AnyAtom>()
+  let savedGetAtomState: StoreArgs[0]
+  const derivedStore = store.unstable_derive(
+    (getAtomState, atomRead, atomWrite, atomOnMount) => {
+      savedGetAtomState = getAtomState
+      return [
+        getAtomState,
+        atomRead,
+        atomWrite,
+        (atom, setAtom) => {
+          try {
+            const onUnmount = atomOnMount(atom, setAtom)
+            return () => {
+              debugMountedAtoms.delete(atom)
+              onUnmount?.()
             }
+          } finally {
+            debugMountedAtoms.add(atom)
           }
+        },
+      ]
+    },
+  )
+  const devStore: DevStoreRev4 = {
+    // store dev methods (these are tentative and subject to change without notice)
+    dev4_get_internal_weak_map: () => ({
+      get: (atom) => {
+        const atomState = savedGetAtomState(atom)
+        if (atomState.n === 0) {
+          // for backward compatibility
+          return undefined
         }
-        flushPending(pending)
+        return atomState
       },
-    }
-    Object.assign(store, devStore)
+    }),
+    dev4_get_mounted_atoms: () => debugMountedAtoms,
+    dev4_restore_atoms: (values) => {
+      throw new Error('TODO: not implemented yet' || values)
```

Comment:
https://github.com/pmndrs/jotai/pull/2827 introduces batch inside the write function only.
So after #2827 lands, it could look something like:
```ts
{
  ...
  dev4_restore_atoms: (values) => {
    store.set(atom(null, (get, set) => {
      values.forEach(([a, v]) => set(a, v))
    })
  },
}
```